ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/BlockTree/Tests.hs

```diff
@@ -0,0 +1,187 @@
+      , testProperty "imagesAreNonempty" $ forAll (genTestBlockTree branchFactor) $
+          prop_deforestBlockTree_imagesAreNonempty
+      , testProperty "allShareTrunkAnchor" $ forAll (genTestBlockTree branchFactor) $
+          prop_deforestBlockTree_allShareTrunkAnchor
+      , testProperty "fullBranchesAreBranches" $ forAll (genTestBlockTree branchFactor) $
+          prop_deforestBlockTree_fullBranchesAreBranches
+      , testProperty "everyHeaderHashIsInTheMap" $ forAll (genTestBlockTree branchFactor) $
+          prop_deforestBlockTree_everyHeaderHashIsInTheMap
+      , testProperty "prefixMaximalPrefixesAreBranches" $ forAll (genTestBlockTree branchFactor) $
+          prop_deforestBlockTree_prefixMaximalPrefixesAreBranches
+      ]
+    ]
+
+
+
+-- | The nonempty prefixes of an AnchoredFragment are in fact prefixes.
+prop_nonemptyPrefixesOf_nonemptyPrefixesArePrefixes
+  :: (Eq blk, HasHeader blk) => AF.AnchoredFragment blk -> QC.Property
+prop_nonemptyPrefixesOf_nonemptyPrefixesArePrefixes fragment =
+  QC.property . all (flip AF.isPrefixOf fragment) . nonemptyPrefixesOf $ fragment
+
+-- | The nonempty prefixes of an AnchoredFragment are in fact nonempty.
+prop_nonemptyPrefixesOf_nonemptyPrefixesAreNonempty
+  :: (Eq blk, HasHeader blk) => AF.AnchoredFragment blk -> QC.Property
+prop_nonemptyPrefixesOf_nonemptyPrefixesAreNonempty fragment =
+  QC.property . all (not . AF.null) . nonemptyPrefixesOf $ fragment
+
+-- | The nonempty prefixes of an AnchoredFragment are unique.
+prop_nonemptyPrefixesOf_nonemptyPrefixesAreUnique
+  :: forall blk. (Eq blk, HasHeader blk) => AF.AnchoredFragment blk -> QC.Property
+prop_nonemptyPrefixesOf_nonemptyPrefixesAreUnique fragment =
+  let
+    isNew :: forall u. (Eq u) => u -> ([u], Bool) -> ([u], Bool)
+    isNew u (seen, restOk) =
+      if not (elem u seen) && restOk
+        then (u:seen, True)
+        else (seen, False)
+  in QC.property . snd . foldr isNew ([], True) . nonemptyPrefixesOf $ fragment
+
+-- | All the nonempty prefixes should share the original fragment's anchor.
+prop_nonemptyPrefixesOf_allShareInputAnchor
+  :: (HasHeader blk) => AF.AnchoredFragment blk -> QC.Property
+prop_nonemptyPrefixesOf_allShareInputAnchor originalFragment =
+  let sharesTrunkAnchor thisFragment restOk =
+        (((==) `on` AF.anchor) thisFragment originalFragment) && restOk
+  in QC.property . foldr sharesTrunkAnchor True . nonemptyPrefixesOf $ originalFragment
+
+
+
+-- | The head points of all the branches are distinct.
+-- (Points uniquely determine positions in the tree.)
+prop_deforestBlockTree_headPointsAreDistinct
+  :: (HasHeader blk) => BlockTree blk -> QC.Property
+prop_deforestBlockTree_headPointsAreDistinct =
+  let headPointsDistinct thisChain (restOk, seenSoFar) =
+        let hP = AF.headPoint thisChain
+        in ((not $ elem hP seenSoFar) && restOk, hP : seenSoFar)
+  in QC.property . fst . foldr headPointsDistinct (True, mempty) . deforestBlockTree
```

Comment:
Accidentally O(n^2). As a good rule of thumb, `elem` is never the function you want